projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/spm/spm_core.cpp

```diff
@@ -0,0 +1,361 @@
+// MIT License
+//
+// Copyright (c) 2023-2025 Advanced Micro Devices, Inc. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "lib/rocprofiler-sdk/spm/spm_core.hpp"
+#include "lib/common/container/stable_vector.hpp"
+#include "lib/common/utility.hpp"
+#include "lib/rocprofiler-sdk/buffer.hpp"
+#include "lib/rocprofiler-sdk/context/context.hpp"
+#include "lib/rocprofiler-sdk/counters/metrics.hpp"
+#include "lib/rocprofiler-sdk/hsa/queue_controller.hpp"
+#include "lib/rocprofiler-sdk/internal_threading.hpp"
+#include "lib/rocprofiler-sdk/registration.hpp"
+#include "lib/rocprofiler-sdk/spm/spm_dispatch_handlers.hpp"
+
+#include <hsa/hsa_api_trace.h>
+#include <rocprofiler-sdk/fwd.h>
+#include <rocprofiler-sdk/intercept_table.h>
+#include <rocprofiler-sdk/rocprofiler.h>
+
+#include <atomic>
+#include <cstdint>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#define CHECK_HSA(fn, message)                                                                     \
+    {                                                                                              \
+        auto _status = (fn);                                                                       \
+        if(_status != HSA_STATUS_SUCCESS)                                                          \
+        {                                                                                          \
+            ROCP_ERROR << "HSA Err: " << _status << '\n';                                          \
+            throw std::runtime_error(message);                                                     \
+        }                                                                                          \
+    }
+
+namespace rocprofiler
+{
+namespace SPM
```

Comment:
```suggestion
namespace spm
```
We don't use all uppercase namespaces anywhere else.